x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,488 @@
+        }
+
+        Ok(Self {
+            inner: Arc::new(InnerConnection::for_connected_stream(stream, setup)?),
+            read_lock: Mutex::new(()),
+            write_lock: Mutex::new(()),
+        })
+    }
+
+    /// Establish a new connection on an already connected stream.
+    pub fn for_connected_stream(stream: S, setup: Setup) -> Result<Self, ConnectError> {
+        // Register the stream in the reactor.
+        let stream = WrappedStream {
+            stream: Async::new(stream)?,
+        };
+
+        Ok(Self {
+            inner: Arc::new(InnerConnection::for_connected_stream(stream, setup)?),
+            read_lock: Mutex::new(()),
+            write_lock: Mutex::new(()),
+        })
+    }
+}
+
+impl<S: Stream> RustConnection<S> {
+    /// Read or write asynchronously with the given function.
+    async fn do_io<R, E: AsIoError>(
+        &self,
+        mode: PollMode,
+        mut f: impl FnMut(&Self) -> Result<R, E>,
+    ) -> Result<R, E> {
+        loop {
+            // Try the operation.
+            let res = poll_fn(|cx| {
+                let res = CURRENT_CONTEXT.set(cx.waker(), || f(self));
+                Poll::Ready(res)
+            })
+            .await;
+
+            match res {
+                Err(e) if e.as_io_error().map(|e| e.kind()) == Some(io::ErrorKind::WouldBlock) => {}
+                res => return res,
+            }
+
+            // Wait until the stream is readable or writable.
+            match mode {
+                PollMode::Readable => self
+                    .inner
+                    .stream()
+                    .stream
+                    .readable()
+                    .await
+                    .map_err(E::from_io_error)?,
+                PollMode::Writable => self
+                    .inner
+                    .stream()
+                    .stream
+                    .writable()
+                    .await
+                    .map_err(E::from_io_error)?,
+                PollMode::ReadAndWritable => {
+                    // `or` them together.
+                    let read = self.inner.stream().stream.readable();
+                    let write = self.inner.stream().stream.writable();
+
+                    read.or(write).await.map_err(E::from_io_error)?;
+                }
+            }
+        }
+    }
+}
+
+impl<S: Stream + Send + Sync + 'static> RequestConnection for RustConnection<S> {
+    type Buf = Vec<u8>;
+
+    fn send_request_with_reply<'this, 'bufs, 'sl, 'future, R>(
+        &'this self,
+        _bufs: &'bufs [io::IoSlice<'sl>],
+        _fds: Vec<RawFdContainer>,
+    ) -> Fut<'future, Cookie<'this, Self, R>, ConnectionError>
+    where
+        'this: 'future,
+        'bufs: 'future,
+        'sl: 'future,
+        R: x11rb::x11_utils::TryParse,
+    {
+        Box::pin(async move {
+            // Gain exclusive access to the write end of the stream.
+            let _guard = self.write_lock.lock().await;
```

Comment:
Does this "two mutexes"-approach really work? :thinking: 

From libxcb's source code, I learnt that there was "once upon a time" a deadlock between libX11 and the X11 server: The X11 server stops reading from a client when it has stuff to sent, so the client must always read and write at the same time. Only writing without reading could produce a deadlock.

With your "two mutex" approach... well, there could be only a thread/task trying to write a request, so this task also has to read a request. I guess this means it also has to lock the read mutex? But there could be another task in `wait_for_reply()` that already does the waiting, so in this task waiting for the mutex would produce unwanted blocking.

Oh and it could happen that the write thread is blocked for a while and in the mean time the `wait_for_reply()` finishes and so the reader that was present at the beginning of `send_request` goes away and the write thread has to also "start" reading.

For my async experiments I solved this by spawning an extra task that does all the reading and is always "blocked" in `read()`. That way, nothing else has to worry about reading since it is always already in progress. Of course, this makes the synchronisation between things more complicated and also "proper shutdown" becomes more complicated....